### ISSUE TITLE
suboptimal assembly for Ordering::reverse

### ISSUE THREAD

**Issue by tspiteri** [rust-lang/rust#43133]
For this code:
```rust
use std::cmp::Ordering;
use std::mem;
pub fn reverse(a: Ordering) -> Ordering {
    a.reverse()
}
pub fn neg(a: Ordering) -> Ordering {
    unsafe { mem::transmute(-mem::transmute::<_, i8>(a)) }
}
```
I get the following from `cargo rustc --release -- --emit asm`
```asm
; snip...
; function reverse:
        .cfi_startproc
        movl    %edi, %eax
        negb    %al
        sbbb    %cl, %cl
        cmpb    $-1, %dil
        movb    $1, %al
        je      .LBB0_2
        movl    %ecx, %eax
.LBB0_2:
        retq
; snip...
; function neg:
        .cfi_startproc
        negb    %dil
        movl    %edi, %eax
        retq
; snip...
```
The result is the same, as `Ordering` defines `Less`, `Equal` and `Greater` to be −1, 0 and 1, but the code for `reverse` is much longer, and includes a conditional branch.

**Comment by kennytm** [comment]
LLVM is not optimizing the `match` into a `switch` table?

When matching the enum through `discriminant_value`, the optimizer is able to reduce it into a simple negation.

<details><summary>Long test output</summary>

```rust
#![feature(core_intrinsics)]

use std::intrinsics::{discriminant_value, unreachable};

#[derive(PartialEq, Eq)]
pub enum Ordering2 {
    Less = -1,
    Equal = 0,
    Greater = 1,
}

pub fn reverse2(a: Ordering2) -> Ordering2 {
    match a {
        Ordering2::Less => Ordering2::Greater,
        Ordering2::Equal => Ordering2::Equal,
        Ordering2::Greater => Ordering2::Less,
    }
}

pub fn reverse3(a: Ordering2) -> Ordering2 {
    unsafe {
        match discriminant_value(&a) {
            0xffff_ffff_ffff_ffff => Ordering2::Greater,
            0 => Ordering2::Equal,
            1 => Ordering2::Less,
            _ => unreachable(),
        }
    }
}
```

LLVM:

```llvm
; _1::reverse2
; Function Attrs: norecurse nounwind readnone uwtable
define i8 @_1::reverse2::hbb951fc559ba0b9c(i8) unnamed_addr #0 {
start:
  %not.switch.selectcmp = icmp ne i8 %0, 0
  %switch.select = sext i1 %not.switch.selectcmp to i8
  %switch.selectcmp1 = icmp eq i8 %0, -1
  %switch.select2 = select i1 %switch.selectcmp1, i8 1, i8 %switch.select
  ret i8 %switch.select2
}

; _1::reverse3
; Function Attrs: norecurse nounwind readnone uwtable
define i8 @_1::reverse3::hf325c37bc61a71c6(i8) unnamed_addr #0 {
start:
  %1 = zext i8 %0 to i64
  %switch.tableidx = add nuw nsw i64 %1, 1
  %switch.idx.cast = trunc i64 %switch.tableidx to i8
  %switch.offset = sub i8 1, %switch.idx.cast
  ret i8 %switch.offset
}
```

x86_64 (`rustc --target=x86_64-apple-darwin -Copt-level=1 --emit=asm --crate-type=rlib 1.rs`):

```asm
__1::reverse2::hbb951fc559ba0b9c:
	pushq	%rbp
	movq	%rsp, %rbp
	movl	%edi, %eax
	negb	%al
	sbbb	%cl, %cl
	cmpb	$-1, %dil
	movb	$1, %al
	je	LBB0_2
	movl	%ecx, %eax
LBB0_2:
	popq	%rbp
	retq

__1::reverse3::hf325c37bc61a71c6:
	pushq	%rbp
	movq	%rsp, %rbp
	negb	%dil
	movl	%edi, %eax
	popq	%rbp
	retq
```

i686:

```asm
__1::reverse2::hbb951fc559ba0b9c:
	movb	4(%esp), %al
	movl	%eax, %ecx
	negb	%cl
	sbbb	%cl, %cl
	cmpb	$-1, %al
	movb	$1, %al
	je	LBB0_2
	movl	%ecx, %eax
LBB0_2:
	retl

__1::reverse3::hf325c37bc61a71c6:
	xorl	%eax, %eax
	subb	4(%esp), %al
	retl
```

ARM:

```asm
_1::reverse2::hbb951fc559ba0b9c:
	ands	r1, r0, #255
	mov	r0, r1
	mvnne	r0, #0
	cmp	r1, #255
	moveq	r0, #1
	bx	lr

_1::reverse3::hf325c37bc61a71c6:
	rsb	r0, r0, #0
	bx	lr
```

AArch64:

```asm
__1::reverse2::hbb951fc559ba0b9c:
	and	w8, w0, #0xff
	tst	 w0, #0xff
	csetm	 w9, ne
	cmp		w8, #255
	csinc	w0, w9, wzr, ne
	ret

__1::reverse3::hf325c37bc61a71c6:
	neg	 w0, w0
	ret
```

</details>

**Comment by tspiteri** [comment]
This appears to have been fixed since 1.23.0.
